docs/internet-identity-spec.adoc

```diff
@@ -170,9 +173,20 @@ The client application frontend should support delegation chains of length more
 The Internet Identity frontend will use `event.origin` as the “Frontend URL” to base the user identity on. This includes protocol, full hostname and port. This means
 
 * Changing protocol, hostname (including subdomains) or port will invalidate all user identities.
+** However, multiple different frontend URLs can be mapped back to the canonical frontend URL, see <<alternative-frontend-origins>>.
 * The frontend application must never allow any untrusted JavaScript code to be executed, on any page on that hostname. Be careful when implementing a JavaScript playground on the Internet Computer.
 --
 
+[#alternative-frontend-origins]
+=== Alternative Frontend Origins
+To allow flexibility regarding the canister frontend URL, the client may choose to provide the canonical canister frontend URL (`https://<canister_id>.ic0.app` or `https://<canister_id>.raw.ic0.app`) as the `derivationOrigin` (see <<client-auth-protocol>>). This means that Internet Identity will issue the same principals to the frontend (which uses a different origin) as it would if it were using one of the canonical URLs.
+
+*Note:* `https://<canister_id>.ic0.app` and `https://<canister_id>.raw.ic0.app` do _not_ issue the same principals by default. However, this feature can also be used to map `https://<canister_id>.raw.ic0.app` to `https://<canister_id>.ic0.app` principals or vice versa.
+
+In order for Internet Identity to accept the `derivationOrigin` the corresponding canister must list the frontend origin in the https://internetcomputer.org/docs/current/references/ic-interface-spec/#state-tree-canister-information[_public_ canister metadata] with name `ii_alternative_origins`. The value of this metadata must be candid encoded `(Vec<Text>)`, where each element is an allowed alternative frontend URL.
```

Comment:
Ah, that feature comes again.

Abstractly, we need a trusted way for the II frontend to learn from derivationOrigin  which alternative origins are allowed.

I would strongly advise against using a IC-specific meachanism to implement this mapping. We deliberately designed the Internet Identity protocol so that it is completely oblivious as to where a certain client frontend is hosted, whether on the Internet Identity. Instead, we relied only on Web and HTTP related technology. Right now, I can host a webpage on `https://foo.nomeata.de`, and allow users to log in using II, and if I later move that page onto the Internet Computer (once it supports custom URLs, which I fail to understand why it still doesn’t), I can take my users with me.

I think throwing that property out would be an unnecessary layer violation that bars certain applications.

So what would be an alternative? If all we have is an URL, we have to use HTTP? I’d go for a https://en.wikipedia.org/wiki/Well-known_URI at the derivationOrigin. For non-raw canister URLs that will secure thanks to the service workers. And raw urls really shouldn’t be used, just like http:// URLs shouldn’t…

Reply:
How does an app developer verify ownership of any of their `alternative-frontend-origins`? Seems like they could be dishonest about which origins they own if they're allowed to add any in their metadata without verification.

Reply:
I think that is not a problem: If nomeata.de lists dostro.com as their alternative-frontend-origins, then that just means that nomeata.de announces it trusts dostro.com.  This on its own does affect dostro.com’s security at all!

Only when the frontend served from dostro.com also sets the derivationOrigin to nomeata.de something will happen. So there is opt-in from both sides.

Reply:
I missed the double opt-in bit :) thank you for the clarification!

Reply:
@nomeata: Thanks for the input! I agree that it would be good to keep II independent of canister infrastructure. I was initially worried about using HTTP requests because currently headers and status codes are not covered by certification. Thus any malicious replica or boundary node could redirect to a malicious endpoint not covered by certification. But the `fetch` API allows to suppress redirects, which would make it secure against this attack.

So the following caveats would apply:
* The URL _must not_ issue a redirect. This breaks with standard semantics of HTTP requests, but other precedents exist: e.g. service workers also cannot be installed from an URL that returns a redirect.
* Canisters that have used a `.raw` domain so far will be stuck in a worse place until we enforce certification on `.raw`, because any malicious replica or boundary node could issue a valid mapping for the given `.raw` well-known URL. Such applications do exist (e.g. DSCVR).
* Apparently, asset certification is hard to do in Motoko at the moment. This seems like a tooling issue, but I'm not sure. @roman-kashitsyn : Could you elaborate on this point?

@bjoernek: What do you think? Given the points above, should we move to using a well-known HTTP URL?

Reply:
Switched spec to use HTTP requests. Pending security review and confirmation by the SDK team that the CORS headers can be implemented within a reasonable timeframe.

Reply:
Update regarding CORS headers: The SDK team should be able to provide this within the required timeframe.